src/legacy/server/logging/rotate/log_rotator.ts

```diff
@@ -0,0 +1,364 @@
+    }
+  }
+
+  _createExitListener() {
+    process.on('exit', this.stop);
+  }
+
+  _deleteExitListener() {
+    process.removeListener('exit', this.stop);
+  }
+
+  async _getLogFileSizeAndCreateIfNeeded() {
+    try {
+      const logFileStats = await statAsync(this.logFilePath);
+      return logFileStats.size;
+    } catch {
+      // touch the file to make the watcher being able to register
+      // change events
+      await writeFileAsync(this.logFilePath, '');
+      return 0;
+    }
+  }
+
+  async _callRotateOnStartup() {
+    this.logFileSize = await this._getLogFileSizeAndCreateIfNeeded();
+    await this._rotate();
+  }
+
+  _shouldRotate() {
+    // should rotate evaluation
+    // 1. should rotate if current log size exceeds
+    //    the defined one on everyBytes
+    // 2. should not rotate if is already rotating or if any
+    //    of the conditions on 1. do not apply
+    if (this.isRotating) {
+      return false;
+    }
+
+    return this.logFileSize >= this.everyBytes;
+  }
+
+  async _rotate() {
+    if (!this._shouldRotate()) {
+      return;
+    }
+
+    await this._rotateNow();
+  }
+
+  async _rotateNow() {
+    // rotate process
+    // 1. get rotated files metadata (list of log rotated files present on the log folder, numerical sorted)
+    // 2. delete last file
+    // 3. rename all files to the correct index +1
+    // 4. rename + compress current log into 1
+    // 5. send SIGHUP to reload log config
+
+    // rotate process is starting
+    this.isRotating = true;
+
+    // get rotated files metadata
+    const foundRotatedFiles = await this._readRotatedFilesMetadata();
+
+    // delete number of rotated files exceeding the keepFiles limit setting
+    const rotatedFiles: string[] = await this._deleteFoundRotatedFilesAboveKeepFilesLimit(
+      foundRotatedFiles
+    );
+
+    // delete last file
+    await this._deleteLastRotatedFile(rotatedFiles);
+
+    // rename all files to correct index + 1
+    // and normalize
+    await this._renameRotatedFilesByOne(rotatedFiles);
+
+    // rename current log into 0
+    await this._rotateCurrentLogFile();
+
+    // send SIGHUP to reload log configuration
+    this._sendReloadLogConfigSignal();
+
+    // Reset log file size
+    this.logFileSize = 0;
+
+    // rotate process is finished
+    this.isRotating = false;
+  }
+
+  async _readRotatedFilesMetadata() {
+    const logFileBaseName = basename(this.logFilePath);
+    const logFilesFolder = dirname(this.logFilePath);
+    const foundLogFiles: string[] = await readdirAsync(logFilesFolder);
+
+    return (
+      foundLogFiles
+        .filter(file => new RegExp(`${logFileBaseName}\\.\\d`).test(file))
+        // we use .slice(-1) here in order to retrieve the last number match in the read filenames
+        // @ts-ignore here as the warning is not too important
+        .sort((a, b) => Number(a.match(/(\d+)/g).slice(-1)) - Number(b.match(/(\d+)/g).slice(-1)))
```

Comment:
Well. I'd prefer to use  `!`  for ts error than mute the check for the whole line.

Reply:
👍